File: src-tauri/src/db/builders/node.rs
//! Node Builder 模式
//!
//! 提供链式 API 简化节点创建

use sqlx::types::Json;
use uuid::Uuid;

use crate::db::{
    insert_node, DbPool, NewNode, NodeType, ResourceEmbeddingStatus, ResourceProcessingStage,
    ResourceSubtype, ReviewStatus, SourceMeta, TaskPriority, TaskStatus,
};

pub struct NodeBuilder {
    uuid: String,
    user_id: i64,
    title: String,
    summary: Option<String>,
    node_type: NodeType,
    task_status: Option<TaskStatus>,
    priority: Option<TaskPriority>,
    due_date: Option<String>,
    done_date: Option<String>,
    file_hash: Option<String>,
    file_path: Option<String>,
    file_content: Option<String>,
    user_note: Option<String>,
    resource_subtype: Option<ResourceSubtype>,
    source_meta: Option<SourceMeta>,
    embedded_hash: Option<String>,
    processing_hash: Option<String>,
    embedding_status: ResourceEmbeddingStatus,
    last_embedding_at: Option<String>,
    last_embedding_error: Option<String>,
    processing_stage: ResourceProcessingStage,
    review_status: ReviewStatus,
}

impl NodeBuilder {
    /// 创建资源节点 Builder
    pub fn resource() -> Self {
        Self::new(NodeType::Resource)
            .with_review_status(ReviewStatus::Unreviewed)
    }

    /// 创建任务节点 Builder
    pub fn task() -> Self {
        Self::new(NodeType::Task)
            .with_task_status(TaskStatus::Todo)
            .with_priority(TaskPriority::Medium)
            .with_review_status(ReviewStatus::Reviewed)
    }

    /// 创建主题节点 Builder
    pub fn topic() -> Self {
        Self::new(NodeType::Topic)
            .with_review_status(ReviewStatus::Reviewed)
    }

    fn new(node_type: NodeType) -> Self {
        Self {
            uuid: Uuid::new_v4().to_string(),
            user_id: 1,
            title: String::new(),
            summary: None,
            node_type,
            task_status: None,
            priority: None,
            due_date: None,
            done_date: None,
            file_hash: None,
            file_path: None,
            file_content: None,
            user_note: None,
            resource_subtype: None,
            source_meta: None,
            embedded_hash: None,
            processing_hash: None,
            embedding_status: ResourceEmbeddingStatus::Pending,
            last_embedding_at: None,
            last_embedding_error: None,
            processing_stage: ResourceProcessingStage::Todo,
            review_status: ReviewStatus::Unreviewed,
        }
    }

    // ========== 基础字段 ==========

    /// 设置标题
    pub fn title(mut self, title: impl Into<String>) -> Self {
        self.title = title.into();
        self
    }

    /// 设置摘要
    pub fn summary(mut self, summary: Option<impl Into<String>>) -> Self {
        self.summary = summary.map(|s| s.into());
        self
    }

    /// 设置用户 ID
    pub fn user_id(mut self, user_id: i64) -> Self {
        self.user_id = user_id;
        self
    }

    /// 设置 UUID（通常由 Builder 自动生成）
    pub fn uuid(mut self, uuid: impl Into<String>) -> Self {
        self.uuid = uuid.into();
        self
    }

    // ========== 任务相关字段 ==========

    /// 设置任务状态
    pub fn task_status(mut self, status: Option<TaskStatus>) -> Self {
        self.task_status = status;
        self
    }

    fn with_task_status(mut self, status: TaskStatus) -> Self {
        self.task_status = Some(status);
        self
    }

    /// 设置优先级
    pub fn priority(mut self, priority: Option<TaskPriority>) -> Self {
        self.priority = priority;
        self
    }

    fn with_priority(mut self, priority: TaskPriority) -> Self {
        self.priority = Some(priority);
        self
    }

    /// 设置截止日期
    pub fn due_date(mut self, due_date: Option<impl Into<String>>) -> Self {
        self.due_date = due_date.map(|s| s.into());
        self
    }

    /// 设置完成日期
    pub fn done_date(mut self, done_date: Option<impl Into<String>>) -> Self {
        self.done_date = done_date.map(|s| s.into());
        self
    }

    // ========== 资源相关字段 ==========

    /// 设置文件哈希
    pub fn file_hash(mut self, hash: Option<impl Into<String>>) -> Self {
        self.file_hash = hash.map(|s| s.into());
        self
    }

    /// 设置文件路径
    pub fn file_path(mut self, path: Option<impl Into<String>>) -> Self {
        self.file_path = path.map(|s| s.into());
        self
    }

    /// 设置文件内容
    pub fn file_content(mut self, content: Option<impl Into<String>>) -> Self {
        self.file_content = content.map(|s| s.into());
        self
    }

    /// 设置用户备注
    pub fn user_note(mut self, note: Option<impl Into<String>>) -> Self {
        self.user_note = note.map(|s| s.into());
        self
    }

    /// 设置资源子类型
    pub fn resource_subtype(mut self, subtype: Option<ResourceSubtype>) -> Self {
        self.resource_subtype = subtype;
        self
    }

    /// 设置资源来源元数据
    pub fn source_meta(mut self, meta: Option<SourceMeta>) -> Self {
        self.source_meta = meta;
        self
    }

    // ========== 嵌入相关字段 ==========

    /// 设置已嵌入的内容哈希
    pub fn embedded_hash(mut self, hash: Option<impl Into<String>>) -> Self {
        self.embedded_hash = hash.map(|s| s.into());
        self
    }

    /// 设置正在处理的内容哈希
    pub fn processing_hash(mut self, hash: Option<impl Into<String>>) -> Self {
        self.processing_hash = hash.map(|s| s.into());
        self
    }

    /// 设置嵌入状态
    pub fn embedding_status(mut self, status: ResourceEmbeddingStatus) -> Self {
        self.embedding_status = status;
        self
    }

    /// 设置最后嵌入时间
    pub fn last_embedding_at(mut self, at: Option<impl Into<String>>) -> Self {
        self.last_embedding_at = at.map(|s| s.into());
        self
    }

    /// 设置最后嵌入错误
    pub fn last_embedding_error(mut self, error: Option<impl Into<String>>) -> Self {
        self.last_embedding_error = error.map(|s| s.into());
        self
    }

    /// 设置处理阶段
    pub fn processing_stage(mut self, stage: ResourceProcessingStage) -> Self {
        self.processing_stage = stage;
        self
    }

    // ========== 审核状态 ==========

    /// 设置审核状态
    pub fn review_status(mut self, status: ReviewStatus) -> Self {
        self.review_status = status;
        self
    }

    fn with_review_status(mut self, status: ReviewStatus) -> Self {
        self.review_status = status;
        self
    }

    // ========== 构建方法 ==========

    /// 获取生成的 UUID
    pub fn get_uuid(&self) -> &str {
        &self.uuid
    }

    /// 构建 NewNode（不插入数据库）
    pub fn build(&self) -> NewNode<'_> {
        NewNode {
            uuid: &self.uuid,
            user_id: self.user_id,
            title: &self.title,
            summary: self.summary.as_deref(),
            node_type: self.node_type,
            task_status: self.task_status,
            priority: self.priority,
            due_date: self.due_date.as_deref(),
            done_date: self.done_date.as_deref(),
            file_hash: self.file_hash.as_deref(),
            file_path: self.file_path.as_deref(),
            file_content: self.file_content.as_deref(),
            user_note: self.user_note.as_deref(),
            resource_subtype: self.resource_subtype,
            source_meta: self.source_meta.as_ref().map(|m| Json(m.clone())),
            embedded_hash: self.embedded_hash.as_deref(),
            processing_hash: self.processing_hash.as_deref(),
            embedding_status: self.embedding_status,
            last_embedding_at: self.last_embedding_at.as_deref(),
            last_embedding_error: self.last_embedding_error.as_deref(),
            processing_stage: self.processing_stage,
            review_status: self.review_status,
        }
    }

    /// 插入到数据库并返回 node_id
    pub async fn insert(self, pool: &DbPool) -> Result<i64, sqlx::Error> {
        let new_node = self.build();
        insert_node(pool, new_node).await
    }

    /// 插入到数据库并返回 (node_id, uuid)
    pub async fn insert_with_uuid(self, pool: &DbPool) -> Result<(i64, String), sqlx::Error> {
        let uuid = self.uuid.clone();
        let node_id = self.insert(pool).await?;
        Ok((node_id, uuid))
    }
}

File: src/types/node.ts
import { z } from "zod";

// ============================================
// 枚举类型
// ============================================

export const nodeTypeValues = ["topic", "task", "resource"] as const;
export type NodeType = (typeof nodeTypeValues)[number];

export const taskStatusValues = ["todo", "done", "cancelled"] as const;
export type TaskStatus = (typeof taskStatusValues)[number];

export const taskPriorityValues = ["high", "medium", "low"] as const;
export type TaskPriority = (typeof taskPriorityValues)[number];

export const resourceSubtypeValues = ["text", "image", "pdf", "url", "epub", "other"] as const;
export type ResourceSubtype = (typeof resourceSubtypeValues)[number];

export const reviewStatusValues = ["unreviewed", "reviewed", "rejected"] as const;
export type ReviewStatus = (typeof reviewStatusValues)[number];

export const embeddingStatusValues = ["pending", "synced", "dirty", "error"] as const;
export type EmbeddingStatus = (typeof embeddingStatusValues)[number];

export const processingStageValues = ["todo", "chunking", "embedding", "done"] as const;
export type ProcessingStage = (typeof processingStageValues)[number];

export const relationTypeValues = ["contains", "related_to"] as const;
export type RelationType = (typeof relationTypeValues)[number];

// ============================================
// Zod Schemas
// ============================================

const sqliteDateSchema = z.preprocess((value) => {
  if (value === null || value === undefined) return value;
  if (value instanceof Date) return value;
  if (typeof value === "string") {
    const trimmed = value.trim();
    if (!trimmed) return value;
    const normalized = trimmed.includes("T") ? trimmed : trimmed.replace(" ", "T");
    const parsed = new Date(normalized);
    if (!Number.isNaN(parsed.getTime())) return parsed;
    const parsedUtc = new Date(`${normalized}Z`);
    if (!Number.isNaN(parsedUtc.getTime())) return parsedUtc;
  }
  return value;
}, z.date());

export const sourceMetaSchema = z.object({
  url: z.string().nullable().optional(),
  window_title: z.string().nullable().optional(),
  process_name: z.string().nullable().optional(),
  captured_at: z.string().nullable().optional(),
});

export type SourceMeta = z.infer<typeof sourceMetaSchema>;

export const nodeRecordSchema = z.object({
  node_id: z.number(),
  uuid: z.string(),
  user_id: z.number(),
  title: z.string(),
  summary: z.string().nullable(),
  node_type: z.enum(nodeTypeValues),
  task_status: z.enum(taskStatusValues).nullable(),
  priority: z.enum(taskPriorityValues).nullable(),
  due_date: sqliteDateSchema.nullable(),
  done_date: sqliteDateSchema.nullable(),
  file_hash: z.string().nullable(),
  file_path: z.string().nullable(),
  file_content: z.string().nullable(),
  user_note: z.string().nullable(),
  resource_subtype: z.enum(resourceSubtypeValues).nullable(),
  source_meta: sourceMetaSchema.nullable(),
  embedded_hash: z.string().nullable(),
  processing_hash: z.string().nullable(),
  embedding_status: z.enum(embeddingStatusValues),
  last_embedding_at: z.string().nullable(),
  last_embedding_error: z.string().nullable(),
  processing_stage: z.enum(processingStageValues),
  review_status: z.enum(reviewStatusValues),
  is_pinned: z.boolean(),
  pinned_at: z.string().nullable(),
  created_at: sqliteDateSchema.nullable(),
  updated_at: sqliteDateSchema.nullable(),
  is_deleted: z.boolean(),
  deleted_at: z.string().nullable(),
});

export type NodeRecord = z.infer<typeof nodeRecordSchema>;

export const edgeRecordSchema = z.object({
  edge_id: z.number(),
  source_node_id: z.number(),
  target_node_id: z.number(),
  relation_type: z.enum(relationTypeValues),
  confidence_score: z.number().nullable(),
  is_manual: z.boolean(),
  created_at: sqliteDateSchema.nullable(),
});

export type EdgeRecord = z.infer<typeof edgeRecordSchema>;

export const edgeWithNodeSchema = z.object({
  edge: edgeRecordSchema,
  node: nodeRecordSchema,
});

export type EdgeWithNode = z.infer<typeof edgeWithNodeSchema>;

export const dashboardSchema = z.object({
  tasks: z.array(nodeRecordSchema).default([]),
  resources: z.array(nodeRecordSchema).default([]),
});

export type DashboardData = z.infer<typeof dashboardSchema>;

// ============================================
// Ingest Progress Types
// ============================================

export interface IngestProgress {
  node_id: number;
  status: ProcessingStage;
  percentage?: number;
  error?: string;
}

// ============================================
// Search Types
// ============================================

export interface NodeSearchSummary {
  node_id: number;
  node_type: NodeType;
  title: string;
  summary: string | null;
}

export interface SemanticSearchResult {
  node: NodeSearchSummary;
  score: number;
}
